parley/src/cursor_test.rs

```diff
@@ -0,0 +1,270 @@
+            actual_index = actual.index(),
+            actual_affinity = actual.affinity(),
+            expected_cursor = self.cursor_to_monospace(expected, true),
+            actual_cursor = self.cursor_to_monospace(actual, false),
+        );
+    }
+
+    /// Asserts that the cursor is before the needle.
+    ///
+    /// The needle must be unique in the text to avoid ambiguity.
+    ///
+    /// ### Panics
+    ///
+    /// - If the needle is not found in the text.
+    /// - If the needle is found multiple times in the text.
+    /// - If the cursor has the wrong position.
+    /// - If the cursor doesn't have [`Affinity::Downstream`].
+    #[track_caller]
+    pub fn assert_cursor_is_before(&self, needle: &str, cursor: Cursor) {
+        let index = self.get_unique_index("assert_cursor_is_before", needle);
+
+        let expected_cursor = Cursor::from_byte_index(&self.layout, index, Affinity::Downstream);
+        self.cursor_assertion(expected_cursor, cursor);
+    }
+
+    /// Asserts that the cursor is after the needle.
+    ///
+    /// The needle must be unique in the text to avoid ambiguity.
+    ///
+    /// ### Panics
+    ///
+    /// - If the needle is not found in the text.
+    /// - If the needle is found multiple times in the text.
+    /// - If the cursor has the wrong position.
+    /// - If the cursor doesn't have [`Affinity::Upstream`].
+    #[track_caller]
+    pub fn assert_cursor_is_after(&self, needle: &str, cursor: Cursor) {
+        let index = self.get_unique_index("assert_cursor_is_after", needle);
+        let index = index + needle.len();
+
+        let expected_cursor = Cursor::from_byte_index(&self.layout, index, Affinity::Upstream);
+        self.cursor_assertion(expected_cursor, cursor);
+    }
+
+    /// Compares two cursors and asserts that they are the same.
+    ///
+    /// ### Panics
+    ///
+    /// - If the cursors don't have the same index.
+    /// - If the cursors don't have the same affinity.
+    #[track_caller]
+    pub fn assert_cursor_is(&self, expected: Cursor, cursor: Cursor) {
+        self.cursor_assertion(expected, cursor);
+    }
+
+    /// Prints the TestLayout's text, with the cursor highlighted.
+    ///
+    /// Uses the same format as assertion failures.
+    #[track_caller]
+    pub fn print_cursor(&self, cursor: Cursor) {
+        eprintln!(
+            concat!(
+                "dumping test layout value\n",
+                "      text: '{text}' - ({actual_index}, {actual_affinity})\n",
+                "            {actual_cursor}\n",
+            ),
+            text = self.text,
+            actual_index = cursor.index(),
+            actual_affinity = cursor.affinity(),
+            actual_cursor = self.cursor_to_monospace(cursor, true),
+        );
+    }
+}
+
+// ---
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn cursor_next_visual() {
+        let (mut lcx, mut fcx) = (LayoutContext::new(), FontContext::new());
```

Comment:
Ideally this would use the build-in Roboto font, but of course it doesn't really matter what font is chosen, so long as it is not fundamentally misbehaving

Reply:
Actually of course it matters, for "up/down" button pressing test cases

Reply:
I think we shouldn't worry about up-down testing for now, until we have a UI for tests with multi-line strings.

Like, yes, in theory we could just stick to comparing the numeric values of cursors, but if we want to do that we don't need this PR at all, just assert_eq.